CMakeLists.txt

```diff
@@ -220,51 +221,87 @@ if(BUILD_PYTHON)
   )
 endif()
 
-set(NVFUSER_CODEGEN ${PROJECT_NAME}_codegen)
-add_library(${NVFUSER_CODEGEN} SHARED ${NVFUSER_SRCS})
+# We create both static and shared libraries.
+#
+# Shared libraries are what ships, but a large advantage of static libraries is
+# that symbols are all visible. This allows us to test internal components
+# inside our test or benchmark binaries, even if we do not want said components
+# to be visible to the outside. If we used only shared libraries, then any API
+# we invoked from test binaries would need to be marked as public, even if we
+# did not want to expose it to users.
+#
+# Note technically we create an "OBJECT" library instead of a "STATIC" library.
+# This is just a CMake quirk; an OBJECT library is a better way to implement a
+# "private" (not installed) static library.
+add_library(codegen_internal OBJECT ${NVFUSER_SRCS})
 
 if(NOT MSVC)
-  target_compile_options(${NVFUSER_CODEGEN} PRIVATE
+  target_compile_options(codegen_internal PRIVATE
     -Wall -Wno-unused-function
     -Werror
   )
 endif()
 
-target_compile_definitions(${NVFUSER_CODEGEN} PRIVATE "-DTORCH_CUDA_BUILD_MAIN_LIB")
-target_include_directories(${NVFUSER_CODEGEN} SYSTEM PRIVATE
-  ${CMAKE_SOURCE_DIR}/third_party/flatbuffers/include
+target_compile_definitions(codegen_internal PRIVATE "-DTORCH_CUDA_BUILD_MAIN_LIB")
+target_include_directories(codegen_internal SYSTEM PUBLIC
   ${CMAKE_SOURCE_DIR}/third_party/gloo # TODO: guard this on usage
+  ${CMAKE_SOURCE_DIR}/third_party/flatbuffers/include
+	PRIVATE
   ${CUDA_TOOLKIT_ROOT_DIR}/extras/CUPTI/include
   ${CUDA_INCLUDE_DIRS}
 )
-target_include_directories(${NVFUSER_CODEGEN} PUBLIC
```

Comment:
so this was the part where we were packaging the old header files. looks like they are no longer exposed through adding `codegen_internal` as part of the dependency?

I'm referring to @xwang233 's comment below on the sample cmake example

Reply:
definitions and include directories don't impact packaging in any way. Packaging in CMake comes from the `install` command[s], but note that we also have do packaging things inside our setup.py that cmake knows nothing about.

As I understand it, because codegen_internal is a `PRIVATE` dependency its includes and such do not propagate. It might apply transitively if we use the $<> syntax and add the objects directly to nvfuser_codegen; let me investigate.


- [ ] Investigate using $<> syntax for codegen_internal instead of adding it in the target_link_libraries

Reply:
Yes!

https://github.com/NVIDIA/Fuser/blob/1603f39bab8c1bbe12e38f2b5de53dec3b7cc373/CMakeLists.txt#L240-L243
https://github.com/NVIDIA/Fuser/blob/1603f39bab8c1bbe12e38f2b5de53dec3b7cc373/CMakeLists.txt#L297-L312

This is the part I was talking about. we were doing some questionable things, but somehow did dump everything under `./nvfuser/`, and later the setup.py just blindly package everything in the same hierarchy, so somehow a pip installed nvfuser package happen to still work with the cmake example. (at least it used to).

Reply:
Having said that, we don't have to fix everything in a single PR. Thanks for putting a TODO/TASK for this.

Reply:
Yeah we definitely want to be way more selective about what headers we install. But, also yes: I'll avoid that for this PR.

Reply:
Looked into this. Just using `$<TARGET_OBJECTS:library>` is not enough.

What I found is that CMake does not propagate `PUBLIC` include directories when using the `TARGET_OBJECTS` method of adding a library. Which kind of makes sense, because it's just pulling the built objects and not the library target itself. In our case, it means that our directories that you mention in line 240 do not propagate and thus there are build failures due to the missing include paths.

As per the other thread I'll still look into making the proper directories `PUBLIC` include paths on the installed target.

Reply:
Fixed in latest version.